--- gwibber/microblog/identica.py_ori	2009-03-04 13:24:53.878300484 +0100
+++ gwibber/microblog/identica.py	2009-03-04 13:24:59.609304419 +0100
@@ -122,41 +122,41 @@
 
   def get_messages(self):
     return simplejson.loads(self.connect(
-      "https://identi.ca/api/statuses/friends_timeline.json",
+      "http://identi.ca/api/statuses/friends_timeline.json",
         urllib.urlencode({"count": self.account["receive_count"] or "20"})))
 
   def get_user_messages(self, screen_name):
     try:
       return simplejson.loads(self.connect(
-        "https://identi.ca/api/statuses/user_timeline/"+ screen_name + ".json",
+        "http://identi.ca/api/statuses/user_timeline/"+ screen_name + ".json",
           urllib.urlencode({"count": self.account["receive_count"] or "20"})))
     except Exception:
       profile = [simplejson.loads(self.connect(
-        "https://identi.ca/api/users/show/"+ screen_name +".json"))]
+        "http://identi.ca/api/users/show/"+ screen_name +".json"))]
       return profile
 
   def get_responses(self):
     return simplejson.loads(self.connect(
-      "https://identi.ca/api/statuses/replies.json"))
+      "http://identi.ca/api/statuses/replies.json"))
 
   def get_direct_messages(self):
     return simplejson.loads(self.connect(
-      "https://identi.ca/api/direct_messages.json"))
+      "http://identi.ca/api/direct_messages.json"))
 
   def get_search(self, query):
     return feedparser.parse(urllib2.urlopen(
-      urllib2.Request("https://identi.ca/search/notice/rss",
+      urllib2.Request("http://identi.ca/search/notice/rss",
         urllib.urlencode({"q": query}))))["entries"]
 
   def get_tag(self, query):
     return feedparser.parse(urllib2.urlopen(
-      urllib2.Request("https://identi.ca/index.php",
+      urllib2.Request("http://identi.ca/index.php",
         urllib.urlencode({"action": "tagrss", "tag":
           query}))))["entries"]
 
   def get_group(self, query):
     return feedparser.parse(urllib2.urlopen(
-      urllib2.Request("https://identi.ca/index.php",
+      urllib2.Request("http://identi.ca/index.php",
         urllib.urlencode({"action": "grouprss", "nickname":
           query}))))["entries"]
 
@@ -191,13 +191,13 @@
 
   def send(self, message):
     data = simplejson.loads(self.connect(
-      "https://identi.ca/api/statuses/update.json",
+      "http://identi.ca/api/statuses/update.json",
         urllib.urlencode({"status":message, "source": "Gwibber"})))
     return Message(self, data)
 
   def send_thread(self, message, target):
     data = simplejson.loads(self.connect(
-      "https://identi.ca/api/statuses/update.json",
+      "http://identi.ca/api/statuses/update.json",
         urllib.urlencode({"status":message,
             "in_reply_to_status_id":target.id, "source": "Gwibber"})))
     return Message(self, data)
